File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> combination;
    void solve(vector<int>& candidates, int target,int index){
        if(target == 0){
            ans.push_back(combination);
            return ;
            
        }
        if(index >= candidates.size() || target < 0 )
            return ;
        
        combination.push_back(candidates[index]);
        solve(candidates,target - candidates[index] , index );
        combination.pop_back();
        solve(candidates,target, index + 1);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        solve(candidates,target , 0);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        comb=[]
        def backtrack(target : int , idx : int):
            if target == 0:
                ans.append(comb[:])
                return

            if idx >= len(candidates) or target < 0 : 
                return
            comb.append(candidates[idx])
            backtrack(target - candidates[idx] , idx)
            comb.pop()
            backtrack(target  , idx + 1)
        backtrack(target , 0 )
        return ans
            